Sources/Matrix.swift

```diff
@@ -0,0 +1,370 @@
+ *        list of conditions and the following disclaimer.
+ *
+ *    *    Redistributions in binary form must reproduce the above copyright notice,
+ *        this list of conditions and the following disclaimer in the documentation
+ *        and/or other materials provided with the distribution.
+ *
+ *    *    Neither the name of CosmicMind nor the names of its
+ *        contributors may be used to endorse or promote products derived from
+ *        this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import Foundation
+
+public struct Matrix<T: Arithmetic> {
+    
+    private(set) var rows : Int
+    private(set) var columns : Int
+    
+    internal var data: [T]
+    
+    init(rows: Int, columns: Int, data: [T]) {
+        self.rows = rows
+        self.columns = columns
+        self.data = data
+    }
+    
+    init(rows: Int, columns: Int) {
+        self.rows = rows
+        self.columns = columns
+        self.data = Array.init(repeating: T.zero, count: rows * columns)
+    }
+    
+    /**
+     :name:    size
+     :description:    returns a tuple with number of rows and columns
+     - returns:    (Int, Int)
+     */
+    public var size : (row: Int, column: Int) {
+        return (rows,columns)
+    }
+    
+    subscript(row: Int, column: Int) -> T {
+        get {
+            precondition(validIndex(row: row, column: column), MatrixError.invalid.localizedDescription)
+            return self.data[(row * columns) + column]
+        }
+        set {
+            precondition(validIndex(row: row, column: column ), MatrixError.invalid.localizedDescription)
+            self.data[(row * columns) + column] = newValue
+        }
+    }
+    
+    /**
+     :name:    validIndex
+     :description:    asserts that the (row,col) index is valid
+     - returns:    Bool
+     */
+    private func validIndex(row: Int, column: Int) -> Bool {
+        return row >= 0 && row < rows && column >= 0 && column < columns
+    }
+    
+    /**
+     :name:    column
+     :description:    gets the column using the index
+     - returns:    [Arithmetic]
+     */
+    func row(index: Int) -> [T] {
```

Comment:
I think we should update signature  to `row(at index: Int)`. `row(at: 1)` is more intuitive than `row(index: 1)` in swift context

Reply:
I agree